src/main.rs

```diff
@@ -346,128 +320,111 @@ enum NodeError {
+        let r = accounts
+            .try_fold(HashSet::new(), |mut cc, addr| async move {
+                let _ = cc.insert(AccountAddressEq(addr));
+                Ok(cc)
+            })
+            .await?;
+        *canonical_cache = r;
     }
-    let mut parallel = 1;
+    let mut stream = node.get_finalized_blocks_from(*height).await?;
+    let timeout = std::time::Duration::from_secs(max_behind.into());
     while !stop_flag.load(Ordering::Acquire) {
-        let mut handles = Vec::with_capacity(parallel as usize);
-        for h in u64::from(*height)..u64::from(*height) + parallel {
+        let (error, chunk) = stream
+            .next_chunk_timeout(max_parallel as usize, timeout)
+            .await
+            .map_err(|_| NodeError::Timeout)?;
+        let handles = chunk.into_iter().map(|fb| {
             let mut node = node.clone();
-            handles.push(async move {
-                let height: AbsoluteBlockHeight = h.into();
-                let blocks_at_height = node.get_blocks_at_height(height.into()).await?;
-                if blocks_at_height.len() != 1 {
-                    return Ok(Err(QueryFailure::NoBlocks));
-                }
-                let block = blocks_at_height[0];
-                let info = node.get_block_info(&block).await;
-                let info = match info {
-                    Ok(info) if !info.finalized => {
-                        return Ok(Err(QueryFailure::NotFinalized));
-                    }
-                    Ok(info) => info,
-                    Err(endpoints::QueryError::NotFound) => {
-                        return Ok(Err(QueryFailure::NotFound));
-                    }
-                    Err(endpoints::QueryError::RPCError(e)) => return Err(e.into()),
+            async move {
+                let binfo = node.get_block_info(&fb.block_hash.into()).await?;
+                let events = if binfo.response.transaction_count == 0 {
+                    Vec::new()
+                } else {
+                    node.get_block_transaction_events(&fb.block_hash.into())
+                        .await?
+                        .response
+                        .try_collect()
+                        .await?
                 };
-                let summary = node.get_block_summary(&block).await?;
-                Ok::<_, anyhow::Error>(Ok((info, summary)))
-            });
-        }
-        let mut success = true;
+                let special = node
+                    .get_block_special_events(&fb.block_hash.into())
+                    .await?
+                    .response
+                    .try_collect()
+                    .await?;
+                Ok::<
+                    (BlockInfo, Vec<BlockItemSummary>, Vec<SpecialTransactionOutcome>),
+                    anyhow::Error,
+                >((binfo.response, events, special))
+            }
+        });
         for result in futures::future::join_all(handles).await {
```

Comment:
We could optimize this a bit by processing as soon as something is ready, using either [FuturesOrdered](https://docs.rs/futures/latest/futures/stream/struct.FuturesOrdered.html) or [FuturesUnordered](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html), but maybe out of scope of this PR

Reply:
I changed it now to use futuresordered now, but only replacing the join_all.

The other change would be more complex.